src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h

```diff
@@ -436,6 +676,236 @@ namespace OpenMS
+        // refill spectrum
+        for (int i = 0; i < mz_positions.size(); ++i)
+        {
+          typename MapType::PeakType peak;
+          peak.setMZ(mz_positions[i]);
+          peak.setIntensity(intensities[i]);
+          average_spec.push_back(peak);
+        } 
+        
+        // store spectrum temporarily
+        exp_tmp.addSpectrum(average_spec);
+      }
+      
+      logger.endProgress();
+      
+      // loop over blocks
+      int n(0);
+      //typename MapType::SpectrumType empty_spec;
+      for (AverageBlocks::ConstIterator it = spectra_to_average_over.begin(); it != spectra_to_average_over.end(); ++it)
+      {
+        exp[it->first] = exp_tmp[n];
+        //exp_tmp[n] = empty_spec;
+        ++n;
+      }
+      
+    }
+    
+    /**
+     * @brief average spectra (centroid mode)
+     *
+     * Averages spectra in centroid mode of one MS level in an experiment. The blocks of spectra to be combined
+     * and their relative weights have previously determined.
+     *
+     * @param exp   experimental data to be averaged
+     * @param spectra_to_average_over    mapping of spectral index to set of spectra to average over with corresponding weights
+     * @param ms_level    MS level of spectra to be averaged
+     */
+    template <typename MapType>
+    void averageCentroidSpectra_(MapType & exp, const AverageBlocks & spectra_to_average_over, const UInt ms_level)
+    {
+      MapType exp_tmp;    // temporary experiment for averaged spectra
+      
+      double mz_binning_width(param_.getValue("mz_binning_width"));
+      String mz_binning_unit(param_.getValue("mz_binning_width_unit"));
+      
+      unsigned progress = 0;
+      const ProgressLogger logger;
+      std::stringstream progress_message;
+      progress_message << "averaging centroid spectra of MS level " << ms_level;
+      logger.startProgress(0, spectra_to_average_over.size(), progress_message.str());
+
+      // loop over blocks
+      for (AverageBlocks::ConstIterator it = spectra_to_average_over.begin(); it != spectra_to_average_over.end(); ++it)
+      {
+        logger.setProgress(++progress);
+        
+        // collect peaks from all spectra
+        // loop over spectra in blocks
+        std::vector<std::pair<double,double> > mz_intensity_all;    // m/z positions and peak intensities from all spectra
+        for (std::vector<std::pair<Size, double> >::const_iterator it2 = it->second.begin(); it2 != it->second.end(); ++it2)
+        {
+          // loop over m/z positions
+          for (typename MapType::SpectrumType::ConstIterator it_mz = exp[it2->first].begin(); it_mz < exp[it2->first].end(); ++it_mz)
+          {
+            std::pair<double,double> mz_intensity(it_mz->getMZ(), (it_mz->getIntensity() * it2->second));    // m/z, intensity * weight
+            mz_intensity_all.push_back(mz_intensity);
+          }
+        }
+        
+        sort(mz_intensity_all.begin(), mz_intensity_all.end(), SpectraMerger::comparePair);
```

Comment:
wouldnt it be easier to just require that all input spectra are sorted? as a PRECONDITION?